Create a primer for modules and use statements

lydia-duncan opened the pull request:
Brad realized shortly before the release that we didn't have a primer on
the various new ways to use a module.  Upon further investigation, it
seemed we didn't cover modules in general or private statements either,
so we decided it was best to make a single primer that covered all these
cases.

lydia-duncan commented:
@bradcray, here's a pull request for the module primer, when you get the chance to look at it.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+ * Module Primer
+ *
+ * This primer introduces the concept of modules, a strategy for encapsulating
+ * code for later use.  It covers:
```
"strategy" -> "concept"
"later use" -> "use" or "use by other code" (or just drop "for later use"?)

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+ * code for later use.  It covers:
+ *   how to define a module
+ *   namespace control for a module's symbols
+ *   outside access of a module's symbols
```
"outside" -> "external"?
"of" -> "to"?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+ *   how to define a module
+ *   namespace control for a module's symbols
+ *   outside access of a module's symbols
+ *   namespace control when using a module in an outside context, including:
```
I'm tripping over "outside" again (primarily because I don't think of it as an adjective, and it seems non-technical).  Maybe drop "in an outside context"?  Or maybe even drop this entire bullet on the argument that it's subsumed by the "namespace control" bullet above?  (I realize that you're distinguishing between control within the module and outside the module, but is that distinction important in this overview?)

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+ *     explicit exclusion of symbols
+ *     explicit inclusion of symbols
+ *     renaming included symbols
+ *   outside access of modules in different files
```
"cross-file access to modules"?  "cross-file module access"?  "cross-file module uses"?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+
+/* A module is a grouping of code - each program consists of at least one
+   module, and every symbol is associated with some module.  If all the code of
+   a file is not enclosed in an explicit module, defined using the syntax below,
```
"using the syntax below" -> "using the 'module' keyword"?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+ */
+module modToUse {
+
+  /* In this case, foo is a public global variable which lives in the module
```
I prefer to avoid "global" because I don't think of Chapel as really supporting global variables (i.e., 'foo' is not globally accessible, it's only accessible within modToUse and modules that use it).  "module-scoped"?  "module-level"?

"which" -> "that"

"lives in" -> "defined in" or "defined by"?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+  var bar: int = 2;
+
+  /* baz is a public function which lives in modToUse */
+  proc baz (x, y) {
```
Note that you're switching between // and /* ... */ style comments.

Another "lives in"

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+     but their parent module must take additional steps to access the contents
+     of the nested module.  We will cover how to access the contents of other
+     modules later in this primer.
+  */
```
"known as" -> "called"?
"take advantage of" -> "refer to"
"take additional steps to access" -> "can't directly access"
Maybe find a way to introduce the term "sub-module" here?  e.g., "These submodules can..."

Order-wise, it seems like in a "read primers from top to bottom" world, it might be nice to cover simple cases of module 'use's before getting into nested modules.  In particular, it would avoid the vague forward reference to "additional steps" and make it seem more like the way anybody accesses any module that they can see.  But that rearrangement may be more work than it's worth.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+     modules later in this primer.
+  */
+  module Inner1 {
+    /* The variable foobar is utilizing modToUse's foo and bar variables. */
```
"is utilizing" -> "can reference"?  "references"?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+  }
+
+  /* Note that since there is other code after the end of modToUse's definition,
+     modToUse is a submodule of the overarching file module "modules".
```
I think we should avoid this in the primer if at all possible.  Specifically, it's often a sign of confusion when people use both module declarations and top-level code within a file and, style-wise, I think we encourage people to use explicit module statements when a file defines more than a single file-defined module (in fact, I thought the compiler issued a warning -- or used to? -- when a file contains both top-level code and module statements?)

lydia-duncan replied:
The compiler doesn't issue a warning when a file contains both top-level code and module statements.  There's only a warning when it attempts to use another module that defines a main function, iirc.

Should I then make the entire file a module and drop this note entirely?

bradcray replied:
Oh, I remember the case I was thinking of...  This program generates a warning because a lot of users equate 'use' with '#include' and so think it's OK to put 'use's anywhere:

    use Math;
    module MyProgram {
      writeln("Hello, world!");
    }

Then they got surprised when the code above resulted in a no-op execution (because there was an empty outer module and nothing ever use'd / initialized the inner module).

I think I would probably take the approach of wrapping all the code that exercises the modules that you set up in a sibling module of its own (named something indicative of its role as the driver or the main module or something...) so that the file only contains module definitions.  That avoids needing to talk about the impact of the file-scoped code and the implicit module here (and I think the earlier description you had of filenames being treated as module names if there is no top-level code covers that topic well enough).

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+     just adds unnecessary clutter to the namespace.  To avoid these cases, a
+     symbol can be declared "private" - this means that only code defined within
+     the same scope as the definition of this symbol can access it.
+  */
```
"group of code" reads awkwardly to me.  I'd also avoid the word "blessed" which seems colloquial rather than technical.  But mostly I think we should consider the reasons for wanting a private symbol to be well-understood by anyone reading this and should simply talk in terms of what private does and less about why one might want it.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+  /* Here, hiddenFoo is a private global variable, which is only accessible by
+     symbols defined within modToUse
+  */
+  private var hiddenFoo = false;
```
Could this be moved up among the other module-scoped variables?  It seems more stylistically attractive/normal to put them together, and since they were named as being public, it makes sense to show how private things can be declared in contrast to the public things right there.  It also feels like more of a first-order concept than sub-modules.  Ditto the private function just below (which I'd sort next to the public function(s) above).

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+       private variable hiddenFoo and the private function hiddenBaz.  However,
+       any private symbol defined within Inner2 will not be visible to symbols
+       defined outside of Inner2.
+     */
```
As before, I'd just as soon learn about basic 'use's of modules and such before getting into submodules.
"visible to symbols" -> "visible within scopes"?  (it's not really the symbols that have visibility to other symbols, right?)

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+  // At the moment, private cannot be applied to type definitions; type
+  // aliases, and declarations of enums, records, and classes cannot be declared
+  // private.  Private also cannot be applied to fields or methods yet.
+
```
"At the moment" -> "In the current implementation" or something that talks more in terms of code versions than wall time.  I'd also be OK with leaving this paragraph out for maintenance purposes and/or because it'll become obvious to people quickly once they try it (?) and then they can request it from us showing they care?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+  writeln("Access from outside a module");
+
+  /* Often, it is desirable for the contents of a module to be accessed from
+     outside that context.  There are several strategies for accomplishing this:
```
Isn't it always desirable if the module is anything other than the main module?
"that context" seems like a more vague wording than ideal.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+     outside that context.  There are several strategies for accomplishing this:
+
+     First, if only one symbol defined in the module is desired, and it is
+     desired only once, the symbol can be referenced explicitly - this is done
```
Symbols can be explicitly referenced even if these two conditions aren't true.  The current wording suggests that this is the only condition in which they can be.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+    /*
+      Use statements can be inserted at the global scope, as well as any
+      arbitrary inner scope defined.
+    */
```
Another instance of "global" which I don't think we have.  It also seems a little roundabout to say "they can be used here or [virtually] anywhere else."  Do we need this statement?  Or what would be more precise...  "They can be used in any lexical scope that contains executable code?" (hmm, that's a bit of a mouthful...).

If this comment remains, can it be combined with the following one?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+
+    /* A use statement brings all of the visible symbols of a module into the
+       scope which defines the use statement.  These symbols may then be
+       accessed without the module name prefix.
```
"defines" -> "contains"

We've talked before about avoiding phrases implying that the symbols are injected into the current scope for fear of confusion w.r.t. shadowing and the like.  To that end, could this be rephrased to say something like "...makes all of a module's public symbols visible from within the scope that contains the use statement." to avoid that implication?

lydia-duncan replied:
Is "makes all of the module's visible symbols available to the scope which contains the use statement" acceptable?  Public implies that only private symbols are ignored, when symbols in a nested block statement are also not available.

bradcray replied:
Sure.  In my proposal, I was thinking of symbols in nested block statements (or other inner scopes) as not being the module's symbols (they're that scope's symbols), but I can see that that's vague.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+    var bazBarFoo = baz(bar, foo); // should be 28
+    writeln(bazBarFoo);
+  }
+
```
In one of your early examples (maybe in nested modules?), you had a nice comment about what was being referred to from another module.  Similarly, here, it seems like it could be helpful (instead of or in addition to the expected numeric output) to say something like "bar and foo (and baz?) refer to the symbols in modToUse by virtue of the 'use' statement.  This could also be folded into the previous comment to turn the general statement which characterizes the use into a specific observation about this example.

bradcray replied:
(and I think it would be great if subsequent examples could do similarly)

lydia-duncan replied:
Agreed.  I had placed the value in directly as a means of checking my expectations for the output, but it is useful for the reader to also mention where that value came from

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+     be accessed with a prefix.  The following line would cause an error,
+     because the previous use of the module 'modToUse' occurred within the curly
+     braces above:
+  */
```
"defining" -> "containing"

"the symbols must be accessed with a prefix" -- this doesn't feel quite right to me since the outer scope could also contain a 'use' statement.  I.e., it's not that after leaving a scope, you're required to use fully-qualified names again, it's just that the 'use' only applies to the scope's contents.  So I might drop the first sentence and simply say something like "Since the following line doesn't live within a scope that contains a 'use' of 'modToUse', it would generate an error if uncommented since 'foo' cannot be directly referenced and it doesn't qualify it with a module name." (?)

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+
+    /* Use statements apply to the entire scope in which they are defined.  Even
+       if the use statement occurs after code which would access the symbols
+       without prefix, these accesses are still valid.
```
"would access the symbols without prefix" -> "refers to its symbols" or "directly refers to its symbols"  (and then "accesses" -> "references")

If you wanted, you could also say something like "Like other Chapel forms of introducing symbols into a scope, use statements...." to imply that there's precedent here and that it's not just that uses are weird.

lydia-duncan replied:
In your second paragraph, are you implying that symbols can be accessed before they are declared?  I'm certain that isn't true, and I don't know what other forms you could be referring to . . .

bradcray replied:
Chapel symbol references within a scope will resolve to symbols declared within that scope over symbols in outer scopes regardless of declaration order.  So classes can be defined in any order and refer to one another without a forward reference, e.g. (and arguably functions, though that gets into the overloading issue again).  That said, if the evaluation of the symbol's definition requires executing some code, it doesn't mean that it's legal to refer to the symbol before its use.  Take the following example:

    var x: int = 42;
    {
      writeln(x);
      var x: int = 36;

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+       to the symbols defined at the same level as the use itself.  Thus, if
+       another bar were defined here, and an access was attempted, the compiler
+       would find the bar at this scope, rather than modToUse.bar.
+    */
```
"are considered ... as the use itself." -- for simplicity, what about replacing this with something like "are only considered when symbols that cannot be resolved directly within the local scope."  While I know I often talk about things like injecting a scope outside of the current scope, I think it's confusing to try and describe that way.

Isn't it also the case that statements like this one only apply to symbols that shadow others, like variables and constants, but not to things like functions where overloads at all visible scopes will be considered?  Is there some way we can reflect that here without making things overly complicated?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+
+    use modToUse;
+
+    writeln(bar); // Will output 4.0, not 2
```
Digging further into one of my previous comments, things like "Will output 4.0, not 2" assume the reader will remember that modToUse.bar had the value of 2 (or will be motivated enough to check) whereas a comment like "refers to the local 'bar' variable with value 4.0, not the 'bar' in 'modToUse' would be more illuminating (talks about which symbols are being referenced rather than their values) and rely less on memory of how modToUse was defined.  (alternatively, modToUse could define more memorable symbols like 'pi' or 'legalVotingAge' which would be self-defining, but that requires more effort and arguably still requires a certain amount of memory).

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+       to the symbols defined outside of the scope where the use statement
+       applies.  Thus, if another bar were defined outside of these curly
+       braces, and an access was attempted, the compiler would find the bar from
+       modToUse, rather than the outer bar.
```
Here again, I'd avoid referring to things like "inner scope" and "outside the scope" and simply emphasize that after trying to resolve symbols in a local scope, symbols exposed by the scope's 'use' statements will be considered prior to looking upward into enclosing lexical scopes (so 'use' statements can shadow variables in parent scopes).  My wording is terrible here, but the idea would be to shift from terms that suggest that 'use's somehow wedge in-between existing lexical scopes (which I know I'm guilty of when talking about them... sorry) and to try and speak more in terms of thinking of them as a place to look for symbols between the local scope and its enclosing/parent scopes.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+  {
+    /* And a scope may contain multiple use statements */
+    use modToUse;
+    use AnotherModule;
```
These two examples are nice.  I  might add ", ThirdModule" to the second 'use' statement to drive home that multiple modules and multiple use statements are completely orthogonal.  You could also say "Equivalently, a scope may contain..." since there isn't any semantic difference between one use statement with multiple modules or multiple use statements (right?).  Or that could be made in a separate more explicit (and probably clear) statement.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+
+  {
+    /* In either case, the modules used in this way are considered at the same
+       scope.  This means that if two modules each define a symbol with the same
```
The wording "same scope" feels thematically related to my balking about inner/outer scopes, though a little harder to resolve.  I guess I'm OK with the current wording but wanted to mention it in case this general theme resonates with you and suggests a different approach.  (I was trying to think of ways to say something like "Modules 'use'd within a given scope do not have any specific order relative to one another.  This means that..." -- but that's vague...  but maybe with more work...)

lydia-duncan replied:
Here's my rephrasing: "In either case, the modules used in this way are considered at the same time (after symbols defined at this scope but before symbols defined outside of it) - the ordering within a use statement or of use statements does not affect the precedence of symbols that share a name. This means that if two modules each define a symbol with the same name, and both modules are used at the same scope, attempts to access a symbol by that name will result in a naming conflict:"

bradcray replied:
This seems much better.  I'd suggest considering:
"at the same time" -> "simultaneously" (for brevity) or maybe "in concert" (because it's not really at the exact same time)
"or of use statements" -> "or across multiple use statements" (for clarity and to get more syntactic symmetry across your 'or' by starting each clause with a preposition)
Otherwise, I think this looks great

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+
+    writeln(foo); // Outputs 12
+    /* The following line would fail because both modToUse and Conflict define a
+       bar:
```
"define a bar" -> "define a symbol named bar" or "define a symbol 'bar'"?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+  }
+
+  writeln();
+  writeln("Limiting a use");
```
There was a writeln() like this at the beginning of the file which gave me pause -- this one does as well.  I understand that these are intended to demarcate major sections of the test, but it seems like this would be clearer if done via formatting in comments.  Say:

    // Limiting a use
    // ===========

rather than a writeln() since it's more a meta-comment to the reader than it is anything about how the program executes (i.e., the output is going to be a bunch of numbers and such that won't be very sensitive to which section they were in?).  OTOH, I guess that the writeln()s break up a huge list of indistinguishable numbers into smaller lists of indistinguishable numbers, so make it simpler to associate specific writeln()s with specific output lines?  But if that were a concern, I'd rather decorate the writelns themselves to say things like writeln("fully qualified access: ", modToUse.bar));  rather than writeln(modToUse.bar);

lydia-duncan replied:
The intention was to break up the output so I could more easily tell if a section was behaving as I expected it to.  I was trying to avoid annotating all the output lines to optimize time writing the output versus time verifying the output is correct.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+    use Conflict only other;
+
+    writeln(foo); // Outputs 12
+    writeln(bar); // Outputs 2
```
Here's another case where saying "refers to the copy of bar in modToUse" feels more helpful in terms of understanding what's going on than "outputs 2".  (though again, I think this comment applies to most every reference and I'm just calling out the ones that seem particularly confusing).  Similarly, the "use Conflict only other;" could be further clarified by having a comment say something like "because of this only clause, Confict's 'bar' is not directly accessible in this scope."  Again, this could go into the block comment just above as a means of connecting its general statements to this specific instance.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+
+  {
+    /* If every symbol other than the one which causes a conflict is desired,
+       you can specify the symbols to exclude via an 'except' list.
```
Would it be possible / easy to change either this 'only' or 'except' example to a list rather than a singleton symbol?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+    use modToUse except bar;
+
+    writeln(foo); // Outputs 12
+    writeln(bar); // Outputs 5 after output in Conflict.bar function
```
I didn't notice that set-up to have a paren-less function conflict with a symbol until now.  I worry a bit that exposing paren-less functions (which many users already find confusing) might just muddy the waters here without adding value.  Is there a strong reason to use it rather than simply another variable named bar?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+
+    use color;
+
+    // The use statement allows you to access it without the prefix
```
"it" -- seems vague -> "an enum's symbols" or ...?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+    /* If both symbols which conflict are desired, or if the use causes symbols
+       to be shadowed which are necessary, you can choose to rename a symbol
+       when including it, so long as the new name does not cause any conflicts
+       with other included symbols.
```
It seems like something in here should refer to the 'as' keyword as the means for doing this renaming.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,390 @@
+       as well as the code for accessing anotherHelper.  A summary compilation
+       line will be provided at the end
+    */
+  }
```
I might be tempted to remove everything from the enum example to here, as it feels less about how the language is defined and more about the quirks of our compiler.  E.g., we've talked about really wanting a compiler that will scan all source files in the module search path to find which modules they define without requiring the filename to match or be listed on the command-line, and I think that's the model we'd like for Chapel.  The above is practical info, but feels almost orthogonal to the namespace control concerns that modules strive to provide and that I think of this primer as trying go teach.

bradcray replied:
Actually, that comment applies to the code following this as well... I just stopped reading once I got to that close brace.

lydia-duncan replied:
I included this section because someone (it may have been @noakesmichael?) expressed an interest in having one place where all of this information was defined.  It's fair to say that a primer isn't the place for it, but I'm not sure where a better place would be.  Do you have any suggestions?

bradcray replied:
Maybe the man page?  (if it's not documented there already)

lydia-duncan replied:
I think it's also important to note that uses can be applied to modules defined outside of the current file (as that is the more common utilization of the 'use' keyword)

bradcray replied:
I'd agree with that.  What if, up at the top where we're making general statements about modules (such as how a file with top-level code is treated as a module) we state that modules are often defined across distinct files, typically using a file per module, and that the compiler can be directed to use these modules by naming them on the command-line, via the -M flag, etc., referring to the man page for details?  (I've since checked and the description of the -M flag in the man page covers this pretty well).

bradcray commented on test/release/examples/primers/modulesHelper.chpl:
```diff
@@ -0,0 +1,13 @@
+/* This file is solely intended as a helper file to the modules primer.  It
+   is not intended to be looked at on its own.
+*/
```
Another advantage to removing the material related to files and directories is that it would avoid having to inject the extra files and directories to support this primer...

bradcray commented:
OK, I think I'm done with my review.  The big things that stick out for me are:

* the current discussion focuses almost solely on symbols that shadow one another leaving a big open question about how functions behave w.r.t. overloading and/or shadowing.  I'm not sure the best way to address this, but I think it should be.  One way might be to have some sort of early example about function overloads one of the first times that you show a 'use' (maybe by injecting similarly-named functions that overload nicely into two of your modules?) and then saying something like "the rest of this discussion will focus on symbols that shadow other symbols like variables and types" (?)

* I noted this a few times, but making the comments talk more about how the symbol names are being resolved and less about what their values are would go a long way to making it teach the concepts better, I think.  The .good file will show what the output is -- the comments should speak to what is being referenced and why.

* I think you need to add the new test to the README and Makefile in this directory still?

lydia-duncan commented:
In respect to bullet one - that was laziness on my part, I will take steps to rectify this.  Bullet 2 has already been resolved, and I'll be sure to get to bullet 3

lydia-duncan commented:
Thanks for the in-depth review!

lydia-duncan commented:
Hi @bradcray - the latest commit is an attempt to address your comment about nested modules.  It also actually demonstrates a use of a nested module, which I had neglected to do previously.

There's other changes since the last time you looked as well, but I haven't fulfilled your request for a section on accessing functions, so it's up to you whether you look at them now or when that gets done.

lydia-duncan commented:
Okay, ready to be looked at again!

cassella commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+      // output 2 * (2 + 3) or '10'
+      writeln(baz(3));
+      // Access the function DifferentArguments.baz using the single argument.
+      // Should output 3 - 1, or '1'
```
DifferentArguments.baz(x) returns x - 2.

lydia-duncan replied:
Fixed, thanks for catching!

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+  var bar: int = 2;
+
+  /* A symbol can be declared "private" - this means that only code defined
+     within the same scope as the definition of this symbol can access it.
```
Can't code in nested/inner scopes also refer to private symbols?

lydia-duncan replied:
Right, so long as the nested/inner scope has the private symbol's defining scope as an outer scope.  I'm trying to think of a way to rephrase that - maybe

> only code defined within the same scope as the definition of this symbol (including nested scopes contained within those bounds) can access it.

bradcray replied:
I think that's on the right track.  How about simply "(including nested scopes)" or "(including code in nested scopes)"?  I think that'd be sufficiently clear (and "within those bounds" gives me pause).

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+     within the same scope as the definition of this symbol can access it.
+
+     Here, hiddenFoo is a private global variable, which is only accessible by
+     symbols defined within modToUse
```
"by symbols within" -> "within scopes in" (or "contained in")?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+  }
+
+  /* hiddenBaz is a private function, which is also only accessible by symbols
+     defined within modToUse.
```
ditto the previous comment

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+
+
+module DifferentArguments {
+  /* This function shares a name with a function in modToUse, but requires
```
"requires" -> "takes"

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+    {
+      /* If many of the module's symbols are desired, or the same symbol is
+         desired multiple times, then it is best to utilize what is known as a
+         "use statement".
```
"many" -> "several"
"it is best" -> "you may want to use" or "it can be convenient to use"?  (a Kyle-style mentality might say that fully qualified is always best, period).

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+         executable code.
+
+         A use statement makes all of the module's visible symbols available to
+         the scope which contains the use statement.  These symbols may then be
```
"which contains" -> "that contains" or "containing"

lydia-duncan replied:
Done (I learned something about restrictive clauses today!)

bradcray replied:
I'm not at all expert in them, but have generally been trained to prefer 'that' over 'which', and to think of using 'which' for cases when the clause could be dropped without breaking the completeness of the sentence.  I also use it to avoid duplicate 'that's in a single sentence, or when the clause feels like it ought to be set off by a comma (which maybe is the same criteria as "when the clause could be dropped"...)

lydia-duncan replied:
http://www.kentlaw.edu/academics/lrw/grinker/LwtaThat_Versus_Which.htm, if you're curious :)

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+      */
+      var bazBarFoo = baz(bar, foo);
+      writeln(bazBarFoo);
+    }
```
What do you think about throwing a comment just before this curly bracket saying "Since 'use' statements only affect their containing scope, when we leave a scope like this, we revert to requiring fully-qualified names" or something like that.  This would set up your next comment (without requiring the reader to notice that the curly bracket we just left was the one containing the 'use').

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+         to its symbols, these references are still valid.  This is similar to
+         other Chapel forms of introducing symbols - for instance, class
+         declaration order does not prevent the first declared class from
+         referring to the second.
```
"the first declared class"/"the second" -> "a class declared earlier"/"one declared later"

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+
+         Thus, as in an earlier example, bazBarFoo should store the result of
+         calling modToUse.baz on modToUse.bar and modToUse.foo, which is in this
+         case '28'.
```
"is in this case" -> "is again"?  (it's no different than the previous case... I'm guessing that you'll likely get to other cases)

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+
+    {
+      /* The symbols provided by a use statement are only considered when the
+         name in question cannot be resolved directly within the local scope.
```
This only applies to non-functions, right?  This is why I was thinking it'd be nice to refer to function overloading early and get it out of the way, and then to state that the rest of the primer would only deal with non-functions.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+         Thus, if another bar were defined here, and an access was attempted,
+         the compiler would find the bar at this scope, rather than
+         modToUse.bar.
+      */
```
This is written in the hypothetical, but for this example, it's actually true, not hypothetical isn't it?  So I'd change to more direct statements: "if...were" -> "because...is", "an access was" -> "the compiler will resolve the access", "would" -> "will", etc.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+      use modToUse;
+
+      writeln(bar);
+      // Will output the value of the bar defined in scope (which is '4.0'),
```
"in scope" -> "in this scope"

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+         Thus, if another bar were defined outside of these curly braces, and an
+         access was attempted, the compiler would find the bar from modToUse,
+         rather than the outer bar.  The bar from modToUse is said to be
+         "shadowing" the definition at the outer scope.
```
More hypotheticals here...

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+    }
+
+
+    var bar = false;
```
What do you think about wrapping this 'bar' in yet another curly bracket scope to avoid bleeding it into the entire module (which might cause confusion w.r.t. other chunks of code examples -- even though you've obviously set it up to work as-is -- particularly if the user made local mods to the primer to experiment with different things).

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+
+    {
+      /* Using nested modules combines accessing a module's symbols explicitly
+         and using a top-level module.  If the parent module(s) have not been
```
I don't feel like I understand this first sentence.

lydia-duncan replied:
Sorry, it is a little unclear.  I'm trying to tie back what readers have already learned and show how it applies to the rules for this specific case.  So the second sentence in the paragraph emphasizes that we need to explicitly name the path to the nested module if we haven't used all of its parents (which is the same for any other symbol in the module), and the third sentence reminds us that it is still a module so all the rules for using a module are still applicable.

Would changing it from "combines" to "combines the concepts of" help to clarify that?  The paragraph is really meant to be taken as a whole.

bradcray replied:
I don't know, I found the whole paragraph confusing.  To me nesting modules, and using nested modules, seems like pretty natural composition -- they're just other symbols defined within modules that happen to be modules themselves and all the rules compose the way you'd expect.  Maybe this is simply a matter of trying to say too much in the abstract where a blanket statement ("Modules can be nested and compose as you'd expect") followed by specific examples might be more useful?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+      writeln(baz(3));
+      // Access the function DifferentArguments.baz using the single argument.
+      // Should output 3 - 2, or '1'
+    }
```
As mentioned above, I'd be tempted to move the function resolution example very early such that all preceding definitions of shadowing, conflicts, etc. stand.  By not eliminating functions from consideration early (as you do here), all previous statements about which symbol is being referred to feel qualified to "all non-function symbols" (yet I don't think we'd want to qualify all those statements.  So if we can knock out the difficult/exceptional case first, that'd be ideal.  One way would be to have your first qualified and unqualified examples refer to functions.  Then show this case which deals with overloading.  Then say "for the rest of this primer, we'll focus on non-function symbols."

lydia-duncan replied:
I prefer defining it here (where I've made it clear that multiple modules may be used) so that I don't have to define a function in scope to show how multiple functions with the same name interact.

Additionally, I'd prefer not to introduce a more complicated topic earlier in the primer when the reader doesn't know what rules might be affected by it, as they are more likely to just forget about the additional case because they've been told not to worry about it.  All the new rules after this point (except enums) apply to both types of symbol resolution so it's okay if they forget about how functions differ when reading about the new rules.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+
+    {
+      /* If every symbol other than the one which causes a conflict is desired,
+         you can specify the symbols to exclude via an 'except' list.
```
This is a bit of an overstatement: you can also do this even if only some of the symbols other than the one that causes the conflict is desired.  Rather than stating in terms of desires, how about just stating the facts.  "Using an 'except' list will cause every symbol other than the ones listed to be available."

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+      /* If both symbols which conflict are desired, or if the use causes
+         symbols to be shadowed which are necessary, you can choose to rename a
+         symbol when including it via the 'as' keyword, so long as the new name
+         does not cause any conflicts with other included symbols.
```
"including/included" -> "using/used"?

lydia-duncan replied:
I think I won't make this change, as I've been trying to restrict "using" to apply to only symbols on which a 'use' can operate (and may or may not have been otherwise successful).

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+    writeln("Application to enums");
+
+    {
+      /* 'Use' statements can also be called on enums.  Normally to access one
```
I'd either decapitalize 'Use' (since quotes imply we're setting it off as a keyword) or drop the quotes (as you do in other places) -- either is fine with me.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+      enum color {red, blue, yellow};
+
+      // Normally you must use the enum name as a prefix
+      var aColor = color.blue;
```
Here, you don't actually need to, though, do you, since the following 'use' applies retroactively to this declaration?  Maybe push these examples into curly brackets of their own to separate them?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+      var anotherColor = yellow; // color.yellow
+      writeln(anotherColor);
+
+      // All of the above rules for 'use' statements also apply to 'use's of
```
'use' statements -> 'use's of modules

It's not clear to me when you are and are not quoting 'use' -- I'm not sure I care much since it generally works pretty well either as English ("use statement") or a reference to a keyword ("'use' statement), but I wanted to point it out in case you were interested in unifying (or maybe there is a unifying principle I'm not seeing).

lydia-duncan replied:
I was trying to be unified on treating it as a keyword, but failed.  I did a pass to try for some consistency.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+      use ContainsNesting.Inner1;
+
+      writeln(foobar); // Will output Inner1.foobar, or '14'
+    }
```
What do you think about moving this whole section on nested modules down closer to where the nested module occurs?  It feels like it could come in nearly any order (including after the enum use).  And enums can't be nested, so it'd even make the "all rules above" even more true...  It also seems like one of the most advanced/non-obvious concepts in the primer (so doesn't seem like it needs to come this early compared to some of the much more useful to most cases features).

lydia-duncan replied:
That's a good point, I'll do that.

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+  /* A module defined within another module is called a nested module.  These
+     submodules can refer to symbols defined within their parent module, but
+     their parent module can't directly access the contents of the nested
+     module.
```
"without a use statement or fully qualified name"?

bradcray commented on test/release/examples/primers/modules.chpl:
```diff
@@ -0,0 +1,386 @@
+  }
+}
+
+module ContainsNesting {
```
Could this be called "Outer" for simplicity/symmetry?  (or "OuterNested"?)

bradcray commented:
Still more inline comments than I was hoping -- sorry.  Sorry to let this slide for so long again.  I don't feel a strong need to re-review it a third time -- it's up to you whether or not you want someone else to (if so, Laura might be a good reviewer as a newbie?)

lydia-duncan commented:
I've used most of your comments, and I don't think they resulted in a major enough change to justify an additional reviewer, so I am also satisfied.  Thanks!
